toolchain/mfc/test/test.py

```
@@ -301,4 +329,16 @@ def handle_case(case: TestCase, devices: typing.Set[int]):
             errors.append(f"[bold red]Failed test {case} after {nAttempts} attempt(s).[/bold red]")
             errors.append(f"{exc}")
 
+        # Check if we should abort early due to high failure rate
+        total_completed = nFAIL + nPASS
+        if total_completed >= MIN_CASES_BEFORE_ABORT:
+            failure_rate = nFAIL / total_completed
+            if failure_rate >= FAILURE_RATE_THRESHOLD:
+                cons.print(f"\n[bold red]CRITICAL: {failure_rate*100:.1f}% failure rate detected after {total_completed} tests.[/bold red]")
+                cons.print(f"[bold red]This suggests a systemic issue (bad build, broken environment, etc.)[/bold red]")
```

Ruff (0.14.7)

f-string without any placeholders

Remove extraneous `f` prefix

(F541)